grpcbp/client_middlewares.go

```diff
@@ -101,3 +106,191 @@ func ForwardEdgeContextStreaming(ecImpl ecinterface.Interface) grpc.StreamClient
+
+			rpcCountLabels := prometheus.Labels{
+				localServiceLabel:      serviceSlug,
+				methodLabel:            m,
+				typeLabel:              unary,
+				successLabel:           success,
+				remoteServiceSlugLabel: serverSlug,
+				codeLabel:              status.Code().String(),
+			}
+			clientRPCRequestCounter.With(rpcCountLabels).Inc()
+			clientActiveRequests.With(activeRequestLabels).Dec()
+		}()
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+// PrometheusStreamClientInterceptor is a client-side interceptor that provides Prometheus
+// monitoring for Streaming RPCs.
+//
+// It emits the following metrics:
+//
+// * grpc_client_active_requests gauge with labels:
+//
+//   - grpc_service: the local service slug, serviceSlug arg
+//   - grpc_method: the name of the method called on the gRPC service
+//   - grpc_slug: the name of the remote server the client connects to
+//
+// * grpc_client_latency_seconds histogram and grpc_client_requests_total
+//   counter with labels:
+//
+//   - grpc_service, grpc_method, grpc_slug
+//   - grpc_success: "true" if status is OK, "false" otherwise
+//   - grpc_type: type of request, i.e unary
+//   - grpc_code: the human-readable status code, e.g. OK, Internal, etc
+func PrometheusStreamClientInterceptor(serviceSlug, serverSlug string) grpc.StreamClientInterceptor {
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (_ grpc.ClientStream, err error) {
+
+		clientStream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		monitoredStream := newMonitoredClientStream(
+			clientStream,
+			method,
+			serviceSlug,
+			serverSlug,
+		)
+		return monitoredStream, nil
+	}
+}
+
+// monitoredClientStream wraps grpc.ClientStream to track Prometheus metrics when streaming starts and finishes.
+type monitoredClientStream struct {
+	grpc.ClientStream
+	start       time.Time
+	method      string
+	serviceSlug string
+	serverSlug  string
+}
+
+func newMonitoredClientStream(stream grpc.ClientStream, method, serviceSlug, serverSlug string) *monitoredClientStream {
+	s := &monitoredClientStream{
+		stream,
+		time.Now(),
+		method,
+		serviceSlug,
+		serverSlug,
+	}
+	s.ActiveRequestsInc()
+	return s
+}
+
+func (s *monitoredClientStream) SendMsg(m interface{}) error {
+	return s.ClientStream.SendMsg(m)
+}
+
+// RecvMsg tracks metrics once streaming is complete. Streaming is complete once a non-nil error
+// is returned from grpc.ClientStream. It returns io.EOF when the client has performed a CloseSend. On
+// any non-EOF error, the stream is aborted and the error contains the RPC status.
+func (s *monitoredClientStream) RecvMsg(m interface{}) error {
+	err := s.ClientStream.RecvMsg(m)
+	if err != nil {
+		if err == io.EOF {
+			s.trackMetrics(nil)
+		} else {
+			s.trackMetrics(err)
+		}
+	}
+	return err
+}
```

Comment:
🔕 WDYT about

```suggestion
func (s *monitoredClientStream) RecvMsg(m interface{}) (err error) {
  defer func() {
    if err != nil {
      s.trackMetrics(err)
    }
  }()
	return s.ClientStream.RecvMsg(m)
}
```

and then make trackMetrics understand io.EOF?

Reply:
i like this better. will do.